src/stateStream.ts

```diff
@@ -0,0 +1,82 @@
+import { markName, defaultErrorSubject } from 'rxbeach/internal';
+import { ActionStream } from 'rxbeach/types/helpers';
+import { tag } from 'rxjs-spy/operators';
+import {
+  shareReplay,
+  startWith,
+  switchMap,
+  catchError,
+  skip,
+  filter,
+  tap,
+} from 'rxjs/operators';
+import { combineReducers, RegisteredReducer } from 'rxbeach/reducer';
+import { Subject, of } from 'rxjs';
+
+/**
+ * Create a state stream for a set of reducers.
+ *
+ * The state stream will always emit its last value upon subscription.
+ * On the first subscription to the state stream, it will emit `defaultState`.
+ *
+ * The state stream is not ref counted, which means that the underlying
```

Comment:
That means the stream never stops, continues to calculate state no matter if the state is used or not? Is this meant to be used for view state streams? Seems the wrong pattern to me? A view state stream should always calculate the valid state at the moment of connect?

Reply:
This is a very good point, and warrants a proper discussion. I agree that the "view state stream" should be destroyed once the view is unmounted. If something should persist when the view is unmounted, it should go in another stream.

Reply:
For sure a good point! I think we should have a follow-up discussion on Wednesday.

The reasoning behind this decision was that I would like to try to avoid introducing an [isTemporary](https://github.com/ardoq/ardoq-front/blob/ab3dc69071c59595669173df6ac975d565801497/src/js/ardoq/streams/utils/stateStreamUtils.ts#L31) flag akin to what we have in `createStateStream` in ardoq-front.

From my observations in the codebase (and based on past experience with Redux), I think it's much easier to reason about state streams that aren't bound to a view's lifecycle, but rather the application's lifecycle. 

Three somewhat orthogonal reasons for this:
- **persisting state beyond a view's lifecycle:** a view might at some point unmount unexpectedly, but that doesn't necessarily mean that we wish to reset the state. In many cases, the best user experience is when the view state persist beyond the view's lifecycle. For example when you are editing a gremlin query, but click to open the workspaces module, then go back to editing the gremlin query.
- **reducing state for the view when the view is not mounted:** in some cases, a state stream should be ready to reduce actions before the view is mounted. I.e. we might wish to switch to a view within a different module and set some state. If we need to wait until the view is mounted to set the state, it can lead to some nasty code.
- **consistency and simplicity:** if we can avoid having two types of reduced state streams, one that resets and another that persists throughout the application lifecycle, I think that will simplify application development.

The first problem can be solved using **refCount: false**, while the second problem needs some kind of [global registry that sets up "global"/app-wide state streams](https://github.com/ardoq/ardoq-front/blob/dfe5c7312bd041858fb85a59e800a19647f7eccf/src/js/ardoq/startStateStreams.ts), similar to what is necessary for routines.

As both of you point out, there is also a third use-case: creating reduced state that is bound to a view, which only sets up once the view is mounted and completes once the view unmounts. This is especially true for our visualizations, since we don't want them to spend time building their view models when they are inactive.

However, here we must be careful to not mix two things - **reduced state** and **view model calculation**. 

IMO, view model calculation should obviously be bound to when a view mounts, since it is only necessary for the view when it's active, while **reduced state** can and should be bound to the application's lifecycle.

It's _highly possible_ that there's some important points that I'm missing, which requires us to have reducers that are bound to a view. That's why I think we should discuss this further :)

Reply:
Here's a problem case:
> The view model streams of visualizations should be reduced streams, have the same life-cycle as their view (React) components and be namespaced to each mounted instance. 

In this case the only thing that is in the WIP implementation of `reduceState` that we don't need / don't want is `shareReplay`.

I think one take-away from this is that we don't really have a use case where we need `shareReplay` without `refCount: false`. So maybe we could define a very simple alias for `shareReplay({ refCount: false, bufferSize: 1 })` named something like `persistAndReplay`? Then reduced state streams would need two operators, but we're free to reuse `reduceState` when creating the tooling for view model streams.

Reply:
True, there are some obvious examples from our current views where we wish to have the reset pattern, but still keep reducers around 🤔 I was trying hard to entertain the idea that we can live without them, but that's a bit detached from reality 😄 

However, I don't think it's a good idea to add a `persistAndReplay` operator, since it would most require state stream definitions to be:
```typescript
const state$ = action$.pipe(
   reduceState(),
   persistAndReplay()
);
// instead of simply
const state$ = action$.pipe(
   reduceState()
);
```

IMO, we should either add some option to `reduceState` that allows to configure this behaviour, or create a separate interface, i.e. `reduceViewState`. If we roll with `reduceViewState` it might make sense to also rename `reduceState`.

Reply:
Added another test, which shows an interesting consequence of `({ refCount: false })`:
https://github.com/ardoq/rxbeach/blob/231ad441add1b662c430a893affbe516b8b10e71/src/reduceState.spec.ts#L111-L127

Perhaps, the best solution is actually to roll with `refCount: true` instead. This allows for more fine-grained control as we can distinguish between "permanent" and "temporary" streams by having a registry. 

If you need a "local" stream that resets, you simply avoid registering it in `startGlobalStateStreams` registry.

If you need a "local" stream per view instance, you have a simple method:
```typescript
const createInstanceStream (action$, instanceId) => action$.pipe(
   withNamespace(instanceId),
   reduceState(...)
);
```

Reply:
Finally, I have pushed a commit that switches the implementation to `refCount: true`. I think that's the best way to go :) I've requested a new review